compiler/rustc_mir_transform/src/delay_const_assignments.rs

```diff
@@ -0,0 +1,159 @@
+use crate::MirPass;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_middle::mir::visit::{NonUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use std::collections::hash_map::Entry;
+
+pub struct DelayConstAssignments;
```

Comment:
Bikeshed: this transformation is more commonly called sinking (e.g. [1](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/Sink.cpp) [2](https://github.com/gcc-mirror/gcc/blob/master/gcc/tree-ssa-sink.cc)), perhaps call this `SinkConstAssignments`